### PR TITLE
rustdoc: word wrap CamelCase in the item list table and sidebar

### PR DESCRIPTION
**Pull request by notriddle**
This is an alternative to https://github.com/rust-lang/rust/pull/126209. That is, it fixes the issue that affects the very long type names in https://docs.rs/async-stripe/0.31.0/stripe/index.html#structs.

This is, necessarily, a pile of nasty heuristics. We need to balance a few issues:

- Sometimes, there's no real word break. For example, `BTreeMap` should be `BTree<wbr>Map`, not `B<wbr>Tree<wbr>Map`.

- Sometimes, there's a legit word break, but the name is tiny and the HTML overhead isn't worth it. For example, if we're typesetting `TyCtx`, writing `Ty<wbr>Ctx` would have an HTML overhead of 50%. Line breaking inside it makes no sense.

# Screenshots

| Before | After |
| ------ | ----- |
| image | image

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I'm not a big fan to this approach. Limiting the width of the item name seems better don't you think?

**Comment by BradMarr** [comment]
Wouldn't the <wbr> tag only apply if the width is already limited? Wouldn't this be more of an application layer on top of #126209 instead of an alternative?

File: src/librustdoc/html/escape/tests.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    assert_eq!(&E("First_Second").to_string(), "First<wbr>_Second");
+    assert_eq!(&E("First Second").to_string(), "First Second");
+    assert_eq!(&E("First HSecond").to_string(), "First HSecond");
+    assert_eq!(&E("First HTTPSecond").to_string(), "First HTTP<wbr>Second");
+    assert_eq!(&E("First SecondThird").to_string(), "First Second<wbr>Third");
+    assert_eq!(&E("First<T>_Second").to_string(), "First&lt;<wbr>T&gt;<wbr>_Second");
+    assert_eq!(&E("first_second").to_string(), "first<wbr>_second");
+    assert_eq!(&E("first:second").to_string(), "first:<wbr>second");
+    assert_eq!(&E("first::second").to_string(), "first::<wbr>second");
+    assert_eq!(&E("MY_CONSTANT").to_string(), "MY<wbr>_CONSTANT");
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/escape/tests.rs:26]: Funnily enough, I would have put the break after the `_` and not before. ^^'

Comment (GuillaumeGomez) [reply src/librustdoc/html/escape/tests.rs]: Awesome, thanks!

File: tests/rustdoc-gui/duplicate-macro-reexport.goml [hunk 2]
```diff
@@ -4,11 +4,11 @@ go-to: "file://" + |DOC_PATH| + "/test_docs/macro.a.html"
 wait-for: ".sidebar-elems .macro"
 // Check there is only one macro named "a" listed in the sidebar.
 assert-count: (
-    "//*[@class='sidebar-elems']//*[@class='block macro']//li/a[text()='a']",
+    "//*[@class='sidebar-elems']//*[@class='block macro']//li/a[normalize-space()='a']",
```
Comment (GuillaumeGomez) [inline tests/rustdoc-gui/duplicate-macro-reexport.goml:7]: In a few cases like this one, are there `<wbr>` tags? Seems surprising, no?

Comment (GuillaumeGomez) [reply tests/rustdoc-gui/duplicate-macro-reexport.goml:7]: Ah I see. Please revert these cases then. :)

**Review by camelid — COMMENTED** [review]
Looks good on a first pass. Will do another review when I get a chance.

File: tests/rustdoc/extremely_long_typename.extremely_long_typename.html [hunk 3]
```diff
@@ -0,0 +1 @@
+<li><div class="item-name"><a class="struct" href="struct.CreateSubscriptionPaymentSettingsPaymentMethodOptionsCustomerBalanceBankTransferEuBankTransfer.html" title="struct extremely_long_typename::CreateSubscriptionPaymentSettingsPaymentMethodOptionsCustomerBalanceBankTransferEuBankTransfer">Create<wbr />Subscription<wbr />Payment<wbr />Settings<wbr />Payment<wbr />Method<wbr />Options<wbr />Customer<wbr />Balance<wbr />Bank<wbr />Transfer<wbr />EuBank<wbr />Transfer</a></div></li>
```
Comment (GuillaumeGomez) [inline tests/rustdoc/extremely_long_typename.extremely_long_typename.html:1]: I'm surprised that it generates `<wbr />` whereas we generate `<wbr>` in rustdoc. Dark magic.

Comment (GuillaumeGomez) [reply tests/rustdoc/extremely_long_typename.extremely_long_typename.html:1]: Don't worry, I'm not asking to change it or anything. I was just amused to see this change and wondered why this tag was updated this way. But thanks for confirming my suspicions. ;)

File: tests/rustdoc-gui/type-declation-overflow.goml [hunk 4]
```diff
@@ -70,11 +70,11 @@ go-to: "file://" + |DOC_PATH| + "/lib2/scroll_traits/trait.Iterator.html"
 
 click: ".item-decl .type-contents-toggle"
 assert-property: ("pre.item-decl", {"scrollLeft": 0})
-scroll-to: "//*[@class='rust item-decl']//details/a[text()='String']"
+scroll-to: "//*[@class='rust item-decl']//details/a[normalize-space()='String']"
```
Comment (GuillaumeGomez) [inline tests/rustdoc-gui/type-declation-overflow.goml:73]: Is there supposed to have `wbr` in this case too?

File: tests/rustdoc-gui/source-anchor-scroll.goml [hunk 5]
```diff
@@ -7,11 +7,11 @@ set-window-size: (600, 800)
 // We check that the scroll is at the top first.
 assert-property: ("html", {"scrollTop": "0"})
 
-click: '//a[text() = "barbar" and @href="#5-7"]'
+click: '//a[normalize-space() = "barbar" and @href="#5-7"]'
```
Comment (GuillaumeGomez) [inline tests/rustdoc-gui/source-anchor-scroll.goml:10]: Same question here.

File: tests/rustdoc-gui/sidebar-source-code.goml [hunk 6]
```diff
@@ -65,13 +65,13 @@ go-to: "file://" + |DOC_PATH| + "/src/lib2/another_folder/sub_mod/mod.rs.html"
 click: "#sidebar-button"
 // We wait for the sidebar to be expanded.
 wait-for-css: (".src-sidebar-expanded nav.sidebar", {"width": "300px"})
-assert: "//*[@class='dir-entry' and @open]/*[text()='lib2']"
-assert: "//*[@class='dir-entry' and @open]/*[text()='another_folder']"
-assert: "//*[@class='dir-entry' and @open]/*[text()='sub_mod']"
+assert: "//*[@class='dir-entry' and @open]/*[normalize-space()='lib2']"
```
Comment (GuillaumeGomez) [inline tests/rustdoc-gui/sidebar-source-code.goml:68]: Same question.

File: tests/rustdoc-gui/help-page.goml [hunk 7]
```diff
@@ -66,5 +66,5 @@ go-to: "file://" + |DOC_PATH| + "/test_docs/index.html"
 set-window-size: (1000, 1000) // Popover only appears when the screen width is >700px.
 assert-false: "#help"
 click: "#help-button > a"
-click: "//*[@id='help']//a[text()='the rustdoc book']"
+click: "//*[@id='help']//a[normalize-space()='the rustdoc book']"
```
Comment (GuillaumeGomez) [inline tests/rustdoc-gui/help-page.goml:69]: Is it really worth to add `wbr` for this sentence? Even without it, it works fine (can't get any smaller, after that we switch to mobile and the popover doesn't exist there):

image

File: src/librustdoc/html/templates/sidebar.html [hunk 8]
```diff
@@ -15,7 +15,9 @@ <h2 class="location"> {# #}
             {% for block in blocks %}
                 {% if block.should_render() %}
                     {% if !block.heading.name.is_empty() %}
-                        <h3><a href="#{{block.heading.href|safe}}">{{block.heading.name}}</a></h3>
+                        <h3><a href="#{{block.heading.href|safe}}"> {# #}
+                            {{block.heading.name|wrapped|safe}} {# #}
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/templates/sidebar.html:19]: I don't think it's useful here considering all the longest headings already have whitespace characters, making them split nicely.

Comment (GuillaumeGomez) [reply src/librustdoc/html/templates/sidebar.html:19]: Hum I see. Then nevermind.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors r+ rollup

**Comment by matthiaskrgr** [comment]
@bors r- 

needs refmt

**Pull request merged**